zephyr/include/sof/lib/regions_mm.h

```diff
@@ -10,31 +10,231 @@
+ */
+struct vmh_active_allocation {
+	struct list_item node;
+
+	void *allocation_ptr;
+	uint32_t allocation_size;
+};
+
+/** @struct block_bundle_descriptor
+ *
+ *  @brief This is a struct describing one bundle of blocks
+ *  used as base for allocators blocks.
+ *
+ *  @var block_size size of memory block.
+ *  @var number_of_blocks number of memory blocks.
  */
-extern struct virtual_memory_heap
-	vm_heaps[CONFIG_MP_MAX_NUM_CPUS + VIRTUAL_REGION_COUNT];
+struct block_bundle_descriptor {
+	uint32_t block_size;
+	uint32_t number_of_blocks;
+};
+
+/*
+ * Maybe this heap config should have small bundles first going up to max
+ * size or there should be a sorting mechanism for those ?
+ * Should we assume that bundles are given from smallest to bigest ?
+ */
+
+/** @struct heap_config
+ *
+ *  @brief This is a struct that aggregates block_bundle_descriptor to
+ *  create one cfg that can be passed to heap initiation.
+ *
+ *  @block_bundle_descriptor[] aggregation of bundle descriptors.
+ */
+struct heap_config {
+	struct block_bundle_descriptor block_bundles_table[MAX_MEMORY_ALLOCATORS_COUNT];
+};
+
+struct virtual_memory_heap *vmh_init_heap(const struct heap_config *cfg,
+					  int memory_region_attribute, bool span);
+void *vmh_alloc(struct virtual_memory_heap *heap, uint32_t alloc_size);
+int vmh_free_heap(struct virtual_memory_heap *heap);
+int vmh_free(struct virtual_memory_heap *heap, void *ptr);
+struct virtual_memory_heap *vmh_reconfigure_heap(struct virtual_memory_heap *heap,
+						 struct heap_config *cfg, bool span);
+void vmh_get_default_heap_config(const struct sys_mm_drv_region *region, struct heap_config *cfg);
+struct virtual_memory_heap *vmh_get_heap_by_attribute(uint32_t attr, uint32_t core_id);
+/**
+ * @brief Checks if ptr is in range of given memory range
+ *
+ * @param ptr checked ptr
+ * @param range_start start of checked memory range
+ * @param range_size size of checked memory range.
+ *
+ * @retval false if no overlap detected.
+ * @retval tru if any overlap detected.
+ */
+static inline bool vmh_is_ptr_in_memory_range(void *ptr, void *range_start, size_t range_size)
+{
+	return (uint8_t *)ptr >= (uint8_t *)range_start &&
+	       (uint8_t *)ptr < (uint8_t *)range_start + range_size;
+}
+
+/**
+ * @brief Checks if given memory range overlaps second range.
+ *
+ * @param checked_memory_ptr start of first checked memory range
+ * @param checked_memory_size size of first checked memory range.
+ * @param range_start start of memory range
+ * @param range_size size of range.
+ *
+ * @retval false if no overlap detected.
+ * @retval tru if any overlap detected.
+ */
+static inline bool vmh_do_ranges_overlap(void *checked_memory_ptr, size_t checked_memory_size,
+					 void *range_start, size_t range_size)
+{
+	return vmh_is_ptr_in_memory_range(checked_memory_ptr, range_start, range_size) ||
+	       vmh_is_ptr_in_memory_range((uint8_t *)checked_memory_ptr + checked_memory_size,
+					  range_start, range_size);
```

Comment:
No. `vmh_do_ranges_overlap(0, 16, 8, 8)` will report as non-overlapping which is obviously wrong. Please, get these right. I'm stopping here.

Reply:
Again sorry as per offline conversation I misunderstood the issue.